src/daemon/daemon.cpp

```diff
@@ -1849,99 +1849,109 @@ void mp::Daemon::create_vm(const CreateRequest* request, grpc::ServerWriter<Crea
-        prepare_user_data(user_data_cloud_init_config, vendor_data_cloud_init_config);
-        config->factory->configure(name, meta_data_cloud_init_config, vendor_data_cloud_init_config);
+                return config->factory->prepare_source_image(source_image);
+            };
 
-        std::string mac_addr;
-        while (true)
-        {
-            mac_addr = mp::utils::generate_mac_address();
+            auto fetch_type = config->factory->fetch_type();
+
+            CreateReply reply;
+            reply.set_create_message("Creating " + name);
+            server->Write(reply);
+            auto vm_image = config->vault->fetch_image(fetch_type, query, prepare_action, progress_monitor);
 
-            auto it = allocated_mac_addrs.find(mac_addr);
-            if (it == allocated_mac_addrs.end())
+            reply.set_create_message("Configuring " + name);
+            server->Write(reply);
+            auto vendor_data_cloud_init_config =
+                make_cloud_init_vendor_config(*config->ssh_key_provider, request->time_zone(), config->ssh_username);
+            auto meta_data_cloud_init_config = make_cloud_init_meta_config(name);
+            auto user_data_cloud_init_config = YAML::Load(request->cloud_init_user_data());
+            prepare_user_data(user_data_cloud_init_config, vendor_data_cloud_init_config);
+            config->factory->configure(name, meta_data_cloud_init_config, vendor_data_cloud_init_config);
+
+            std::string mac_addr;
+            while (true)
             {
-                allocated_mac_addrs.insert(mac_addr);
-                break;
+                mac_addr = mp::utils::generate_mac_address();
+
+                auto it = allocated_mac_addrs.find(mac_addr);
+                if (it == allocated_mac_addrs.end())
+                {
+                    allocated_mac_addrs.insert(mac_addr);
+                    break;
+                }
             }
-        }
-        auto vm_desc =
-            to_machine_desc(request, name, checked_args.mem_size, checked_args.disk_space, mac_addr,
-                            config->ssh_username, vm_image, meta_data_cloud_init_config, user_data_cloud_init_config,
-                            vendor_data_cloud_init_config, config->ssh_key_provider.get());
+            auto vm_desc = to_machine_desc(request, name, checked_args.mem_size, checked_args.disk_space, mac_addr,
+                                           config->ssh_username, vm_image, meta_data_cloud_init_config,
+                                           user_data_cloud_init_config, vendor_data_cloud_init_config,
+                                           config->ssh_key_provider.get());
+
+            config->factory->prepare_instance_image(vm_image, vm_desc);
 
-        config->factory->prepare_instance_image(vm_image, vm_desc);
+            emit on_prepare_finished(vm_desc);
 
-        emit on_prepare_finished(vm_desc);
+            QObject::disconnect(connection);
```

Comment:
When you mentioned a junk signal, I got concerned about this code pattern. There is a problem here, and I don't think this fixes it correctly. I made a simple demo to check:
https://git.launchpad.net/~gerboland/+git/multiple-connection-trial/tree/connection.cpp#n10
and if you run it, you'll see there is a problem. 

If 2 launch requests come in near simultaneously, it is possible for there to be 2 signal/slot connections active at the same time, and when the first signal emission happens, the slot will be fired twice.

I think it would be safer to move this lambda into a separate function, and run the connect() in the Daemon constructor. It would avoid trying to do this connect/disconnect dance.

Reply:
Ok, removed this completely by handling the whole thing differently.